### PR TITLE
Configure Services's Server Setting with a Barcode #1373

### PR DESCRIPTION
**Pull request by laddha-adi**
**What has been done to solve this issue?**
I have implemented QR code scanning feature to edit server settings using [this](https://github.com/journeyapps/zxing-android-embedded) library. 
After discussing with @wbrunette, the format for QR code finalized was JSON.

_Examples of how JSON Object for QR code would look like-_ 
**For anonymous access** 
```
{
  "url": "https://yoururl.appspot.com"
}
```
**For username-password access**
```
{
 "url": "https://yoururl.appspot.com",
"username":"user",
"password":"password"
}
```
**What has been done to verify that this works as intended?**
Used QR codes for testing. 

**Are there any risks to merging this code? If so, what are they?**
I don't think so.

### CODE REVIEW

**Comment by jbeorse** [comment]
runtests

**Comment by linl33** [comment]
@laddha-adi thank you for implementing this feature! 
I have a few comments,
1. In the ODK Services app, a device admin could set restrictions on preferences. An admin, for example, might disable modification of server url. You can play with the options in Services > Settings > Enable user restrictions. The QR code function should also respect these restrictions. 

2. There are a few places in your code that have blocks of code that look very similar to this
     ```java
    PropertiesSingleton props = ((IOdkAppPropertiesActivity) this.getActivity()).getProps();
    Map<String, String> properties = new HashMap<String, String>();
    properties.put(CommonToolProperties.KEY_PASSWORD, pw);
    props.setProperties(properties);
    properties.put(CommonToolProperties.KEY_DEFAULT_GROUP, "");
    properties.put(CommonToolProperties.KEY_ROLES_LIST, "");
    properties.put(CommonToolProperties.KEY_USERS_LIST, "");
      ```
      Could you consolidate them? 

3. The way the runtime permission callback is handled is not entirely correct. It does not attempt to show a rationale dialog. You can use `handleRequestPermissionsResult` in `RuntimePermissionUtils` to handle that. 

4. It would be nice if there's a button if Services that could display a QR code with what the device is configured to. This might be better suited a separate pull request though.

**Comment by laddha-adi** [comment]
Hi, @linl33, thanks for your suggestions! I will certainly go through them and would try my best to improve the code. I think I didn't get you on point 4. Can you please elaborate more on it? 
Did you mean that - You wish to have a way to generate QR code on Service app, with configured settings?

**Comment by linl33** [comment]
Yes that is what I was trying to say.

**Comment by laddha-adi** [comment]
We can create another issue for that. I would like to work on it too.

**Comment by laddha-adi** [comment]
Hi, @linl33 I am done with points 1 and 3. Can you please explain a bit more what I am required to do in point 2?

**Comment by linl33** [comment]
```java
      preference.setSummary(newValue.toString());

      PropertiesSingleton props =
              ((IOdkAppPropertiesActivity) ServerSettingsFragment.this.getActivity()).getProps();
      Map<String, String> properties = new HashMap<String, String>();
      properties.put(CommonToolProperties.KEY_SYNC_SERVER_URL, newValue.toString());
      properties.put(CommonToolProperties.KEY_DEFAULT_GROUP, "");
      properties.put(CommonToolProperties.KEY_ROLES_LIST, "");
      properties.put(CommonToolProperties.KEY_USERS_LIST, "");
      props.setProperties(properties);
```
[link](https://github.com/laddha-adi/services/blob/ae48d96b58a2ee3f207dae378fa6ff826d2a5969/services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java#L222-L231)

```java
    int index = ((ListPreference) preference).findIndexOfValue(newValue.toString());
    String entry = (String) ((ListPreference) preference).getEntries()[index];
    ((ListPreference) preference).setSummary(entry);

    PropertiesSingleton props =
            ((IOdkAppPropertiesActivity) ServerSettingsFragment.this.getActivity()).getProps();
    Map<String,String> properties = new HashMap<String,String>();
    properties.put(CommonToolProperties.KEY_AUTHENTICATION_TYPE, newValue.toString());
    properties.put(CommonToolProperties.KEY_DEFAULT_GROUP, "");
    properties.put(CommonToolProperties.KEY_ROLES_LIST, "");
    properties.put(CommonToolProperties.KEY_USERS_LIST, "");
    props.setProperties(properties);
```
[link](https://github.com/laddha-adi/services/blob/ae48d96b58a2ee3f207dae378fa6ff826d2a5969/services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java#L241-L252)

```java
    PropertiesSingleton props = ((IOdkAppPropertiesActivity) this.getActivity()).getProps();
    preference.setSummary((CharSequence) newValue);
    if ( props.containsKey(preference.getKey())) {
      Map<String,String> properties = new HashMap<String,String>();
      properties.put(preference.getKey(), newValue.toString());
      properties.put(CommonToolProperties.KEY_DEFAULT_GROUP, "");
      properties.put(CommonToolProperties.KEY_ROLES_LIST, "");
      properties.put(CommonToolProperties.KEY_USERS_LIST, "");
      props.setProperties(properties);
    } else {
      throw new IllegalStateException("Unexpected case");
    }
```
[link](https://github.com/laddha-adi/services/blob/ae48d96b58a2ee3f207dae378fa6ff826d2a5969/services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java#L305-L316)

These 3 blocks are almost identical. They all, 
1. Update preference summary
2. Update 1 key value pair in PropertiesSingleton
3. Set `KEY_DEFAULT_GROUP`, `KEY_ROLES_LIST` and `KEY_USERS_LIST` in PropertiesSingleton to empty

Since they are so similar, merging them into 1 method (or more) would improve readability.

**Comment by laddha-adi** [comment]
Okay. Btw, this code snippet was from the previous code itself so I preferred not to modify this area. Anyways, I'll do the required changes. Thanks.

**Comment by laddha-adi** [comment]
Hi @linl33, I have done the required changes. Please Review them.

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 1]
```diff
@@ -223,8 +197,66 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+    if (isValid) {
+      preference.setSummary(newValue.toString());
+      updatePreferenceSummary(CommonToolProperties.KEY_SYNC_SERVER_URL, newValue.toString());
+      return true;
+    } else {
+      Toast.makeText(getActivity().getApplicationContext(),
+              R.string.url_error, Toast.LENGTH_SHORT)
+              .show();
+      return false;
+    }  }
```
Comment (clarlars) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: Please put the ending } for the function on a new line for readability.

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 2]
```diff
@@ -273,25 +305,172 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+      updatePreferenceSummary(preference.getKey(), newValue.toString());
     } else {
       throw new IllegalStateException("Unexpected case");
     }
     return true;
   }
 
+  /**
+   * Receives result after QrCode scanning.
+   */
```
Comment (clarlars) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: Please add @Override

Comment (laddha-adi) [reply services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: Done the required changes. Thanks for suggestions!  @clarlars

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 3]
```diff
@@ -273,25 +305,172 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+        return true;
+
+      default:
+        return super.onOptionsItemSelected(item);
+
+    }
+  }
+
+  private void openBarcodeScanner() {
+    FragmentIntentIntegrator integrator = new FragmentIntentIntegrator(this);
```
Comment (clarlars) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: We can remove the FragmentIntentIntegrator class as you can chain the function calls together and use the factory method.  See the example below:

    IntentIntegrator.forFragment(this).
        setDesiredBarcodeFormats(IntentIntegrator.QR_CODE_TYPES).
        setPrompt("Place QR Code inside the rectangle").
        setCameraId(0).setBeepEnabled(true).setBarcodeImageEnabled(false).initiateScan();

Comment (clarlars) [reply services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: If you see any issues with this or any other comment, don't hesitate to let me know.  Thanks!

Comment (laddha-adi) [reply services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: That would work fine I guess. I do the changes.

Comment (laddha-adi) [reply services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: @clarlars,  I have done the required changes. Please review.

**Comment by clarlars** [comment]
Thanks for making the changes!

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 4]
```diff
@@ -50,7 +67,13 @@
   private ListPreference mSignOnCredentialPreference;
   private EditTextPreference mUsernamePreference;
   private ListPreference mSelectedGoogleAccountPreference;
+  PasswordPreferenceScreen passwordScreen;
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: You could remove this variable, it's not used.

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 5]
```diff
@@ -223,8 +196,67 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+    ((ListPreference) preference).setSummary(entry);
+    updatePreferenceSummary(CommonToolProperties.KEY_AUTHENTICATION_TYPE, newValue.toString());
+
+    return true;
+  }
+  @Override
+  public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
+    inflater.inflate(R.menu.server_settings_action_menu, menu);
+  }
+  private void updatePreferenceSummary(String key, String value){
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: This name doesn't reflect what the method does, maybe `updatePropertiesSingleton`?

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 6]
```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+   * Receives result after QrCode scanning.
+   */
+  @Override
   public void onActivityResult(int requestCode, int resultCode, Intent data) {
-    super.onActivityResult(requestCode, resultCode, data);
 
     if (requestCode == RequestCodeConsts.RequestCodes.LAUNCH_CHECKPOINT_RESOLVER ||
         requestCode == RequestCodeConsts.RequestCodes.LAUNCH_CONFLICT_RESOLVER) {
       healthValidator.verifyTableHealth();
+    }
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: You should add a return or else here.

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 7]
```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+     catch (Exception e) {
+       toastString +="Password : not found";
+       Log.i(TAG,"Password not found");
+     }
+
+     Toast.makeText(getActivity(), toastString, Toast.LENGTH_SHORT).show();
+
+    } catch (JSONException e) {
+      Toast.makeText(getActivity(), "Invalid Qr code.", Toast.LENGTH_SHORT).show();
+      Log.i(TAG,"Invalid QR code");
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: This string should be in strings.xml.

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 8]
```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+      default:
+        return super.onOptionsItemSelected(item);
+
+    }
+  }
+
+  private void openBarcodeScanner() {
+    IntentIntegrator.forFragment(this)
+        .setDesiredBarcodeFormats(IntentIntegrator.QR_CODE_TYPES)
+        .setPrompt("Place QR Code inside the rectangle")
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: This string should be in strings.xml

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 9]
```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+                                         @NonNull int[] grantResults) {
+      super.onRequestPermissionsResult(requestCode,permissions,grantResults);
+     if (requestCode == PERMISSION_REQUEST_CAMERA) {
+      // Request for camera permission.
+      if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+        // Permission has been granted. Starting Barcode Scanner.
+         openBarcodeScanner();
+      } else {
+        // Permission request was denied.
+        Toast.makeText(getActivity(), "Permission denied", Toast.LENGTH_SHORT).show();
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: This string should be in strings.xml

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 10]
```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+      }
+    }else{
+       Log.e("here", String.valueOf(requestCode));
+     }
+   }
+
+  private void requestCameraPermission() {
+    Log.i(t, "CAMERA permission has NOT been granted. Requesting permission.");
+    if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: This shouldn't be checked here. It should be in onRequestPermissionsResult

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java [hunk 11]
```diff
@@ -273,24 +305,171 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
...
+        .setBeepEnabled(true)
+        .setBarcodeImageEnabled(false)
+        .initiateScan();
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                         @NonNull int[] grantResults) {
+      super.onRequestPermissionsResult(requestCode,permissions,grantResults);
+     if (requestCode == PERMISSION_REQUEST_CAMERA) {
```
Comment (linl33) [inline services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: We already have a utility method that handles permissions, please use it. 
See BaseLauncherActivity in androidcommon for reference, or TableDisplayActivity in Tables.

**Comment by linl33** [comment]
One more thing, before we can merge your pull request, the debugging Toasts need to be removed.

**Comment by wbrunette** [comment]
runtests

File: services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java (hunk 11 above)
Comment (linl33) [reply services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java]: You're right, I didn't notice that. We request permission from the activity everywhere else. Let's just leave it as is, it's not that big of a problem.

**Comment by laddha-adi** [comment]
Hi, @wbrunette, @linl33  I have made the required changes. Please review them. :)

**Comment by wbrunette** [comment]
runtests

**Pull request merged**